### PR TITLE
Add allowEmptyEndpoints in DynamicEndpointGroup

### LINKED ISSUE #3952: Provide a way to disallow overriding endpoints with empty endpoints for `DynamicEndpointGroup`
A `DynamicEndpointGroup` could be empty because of misconfigured DNS records or consecutive health-check failures.
I think we can provide an option that prevents the existing endpoints from empty endpoints.
For example,
```java
public DynamicEndpointGroup(boolean allowEmptyEndpoints) {
    ...
}

DnsTextEndpointGroup.builder(...)
                    .allowEmptyEndpoint(false);
```
If `allowEmptyEndpoints` is set to false, empty endpoints should not be updated by `setEndpoints` 
https://github.com/line/armeria/blob/39e97e2686c76e8122f2c1ec6057abcd201a485b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java#L160

### PR DESCRIPTION
**Pull request by doppany**
Motivation:

A `DynamicEndpointGroup` could be empty because of misconfigured DNS
records or consecutive health-check failures. We could provide an option
that prevents the existing endpoints from empty endpoints.

Modifications:

- Added more protected constructors to `DynamicEndpointGroup` so that
  a subclass can specify whether to allow empty endpoints.
- Added `DynamicEndpointGroup.builder()` for building a `DynamicEndpointGroup`
  fluently.

Result:

- Closes #3952 
- A user can now configure the `EndpointGroup` implementations based on
  `DynamicEndpointGroup` to allow or disallow empty endpoints.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:78]: You may want to take `EndpointSelectionStrategy` together.

**Comment by ikhoon** [comment]
Could you add `DynamicEndpointGroupBuilder` and make `DnsEndpointGroupBuilder`, `AbstractHealthCheckedEndpointGroupBuilder` and other `EndpointGroupBuilder` extends `DynamicEndpointGroupBuilder`?
```class
abstract class DynamicEndpointGroupBuilder {
    DynamicEndpointGroupBuilder allowEmptyEndpoints(boolean allowEmptyEndpoints) {
        ...
    }
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:170]: 👍

**Comment by trustin** [comment]
@doppany Thanks for your contribution. Could you please sign the CLA and address @ikhoon's comments?

**Comment by doppany** [comment]
@trustin Thank you for reminding me CLA. @ikhoon could you approve that workflow? 
And I also add the condition about `allowEmptyEndpoints` in` removeEndpoint()` as well as in `setEndpoints()`.
+) Sorry for the late commit 😢

**Comment by ikhoon** [comment]
> could you approve that workflow

Done. JFYI, as per GitHub's policy, we have to manually approve to run CI builds. 😄

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:20]:
```suggestion
 * A skeletal builder implementation for an {@link EndpointGroup} builder.
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:29]: - Please add Javadoc for the protected constructors.
- Is allowing empty endpoints the sensible default? If not, I'd recommend removing `= true` as well as the constructors.

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:42]: - Missing Javadoc
- Could be protected final

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:45]:
```suggestion
}
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (doppany) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:29]: @trustin 
> Please add Javadoc for the protected constructors.

Yes, I will. 


> Is allowing empty endpoints the sensible default? If not, I'd recommend removing = true as well as the constructors.

I agree that `allowEmptyEndpoints =true` by default is unnecessary. I will change the logic. thanks 😄

**Comment by trustin** [comment]
@doppany Hey, it's been a while! I see some of the comments in this PR are not addressed yet and there's a merge conflict. Would you mind fixing them once you have a chance? 🙇 

I also find that you added a commit with different username (doyeon-mun). Could you squash all commits and ensure there's only one author in the commit, so that @CLAassistant doesn't complain?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:300]:
```suggestion
    /**
     * Returns whether this {@link EndpointGroup} allows an empty {@link Endpoint} list.
     */
    public boolean allowsEmptyEndpoints() {
        return allowEmptyEndpoints;
    }
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:304]:
```suggestion
    /**
     * Returns a newly created builder.
     */
    public static DynamicEndpointGroupBuilder builder() {
        return new DynamicEndpointGroupBuilder();
    }
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:20]: - Missing Javadoc
- Missing `final`

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:26]:
```suggestion
    private DynamicEndpointGroupBuilder() {}
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:28]: Javadoc

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:38]: Javadoc

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:39]:
```suggestion
        if (selectionStrategy != null) {
```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:48]: How about renaming to `shouldAllowEmptyEndpoints()` so it reads like:

```
if (builder.shouldAllowEmptyEndpoints) {
  ...
}
```

**Comment by trustin** [comment]
We're gonna release a new version soon with this PR merged 🙂 Please don't force-push; I'll add commits to your branch to address the comments by myself to make the release happen sooner.

**Comment by doppany** [comment]
@trustin Oh, I see. Feel free to change it to release sooner. Still, please let me know if you need my hand 🙏

**Comment by doppany** [comment]
@trustin May I fix the failing TC?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java:38]: Should check for nullability since the parameter `selectionStrategy` isn't marked `@Nullable`.
Related to this, what do you think of also giving `selectionStrategy` a default value `weightedRoundRobin ` instead of `null`?

**Comment by jrhee17** [comment]
Can you check the CI failure? The thread dump seems to indicate the following test failure

```
	at com.linecorp.armeria.client.endpoint.healthcheck.HealthCheckedEndpointGroupTest.authTest(HealthCheckedEndpointGroupTest.java:432)
```

ref: https://github.com/line/armeria/runs/5654282909?check_suite_focus=true

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:90]: nit; I'm not sure if there will be a need for this constructor. What do you/other maintainers think?

**Review by jrhee17 — COMMENTED** [review]
Left some minor suggestions, but overall looks really nice!

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:66]: Leaving a simple diff of my idea in case my suggestion was unclear 😅 

```
diff --git a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
index 613426286..4a9b1241f 100644
--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
@@ -16,7 +16,6 @@
 package com.linecorp.armeria.client.endpoint;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
-import static java.util.Objects.requireNonNull;
 
 import java.util.ArrayList;
 import java.util.Collections;
@@ -63,7 +62,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
 
     private final CompletableFuture<List<Endpoint>> initialEndpointsFuture = new EventLoopCheckingFuture<>();
     private final AsyncCloseableSupport closeable = AsyncCloseableSupport.of(this::closeAsync);
-    private boolean allowEmptyEndpoints;
+    private final boolean allowEmptyEndpoints;
 
     /**
      * Creates a new empty instance, using {@link EndpointSelectionStrategy#weightedRoundRobin()}
@@ -79,7 +78,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param selectionStrategy the {@link EndpointSelectionStrategy} of this {@link EndpointGroup}
      */
     public DynamicEndpointGroup(EndpointSelectionStrategy selectionStrategy) {
-        this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
+        this(selectionStrategy, false);
     }
 
     /**
@@ -88,8 +87,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param allowEmptyEndpoints whether to allow an empty {@link Endpoint} list
      */
     protected DynamicEndpointGroup(boolean allowEmptyEndpoints) {
-        this();
-        this.allowEmptyEndpoints = allowEmptyEndpoints;
+        this(EndpointSelectionStrategy.weightedRoundRobin(), allowEmptyEndpoints);
     }
 
     /**
@@ -99,7 +97,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param allowEmptyEndpoints whether to allow an empty {@link Endpoint} list
      */
     protected DynamicEndpointGroup(EndpointSelectionStrategy selectionStrategy, boolean allowEmptyEndpoints) {
-        this(selectionStrategy);
+        this.selectionStrategy = selectionStrategy;
         this.allowEmptyEndpoints = allowEmptyEndpoints;
     }
 
diff --git a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
index a1ee6faa9..d5c6dfafa 100644
--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
@@ -15,7 +15,8 @@
  */
 package com.linecorp.armeria.client.endpoint;
 
-import com.linecorp.armeria.common.annotation.Nullable;
+import static java.util.Objects.requireNonNull;
+
 import com.linecorp.armeria.common.annotation.UnstableApi;
 
 /**
@@ -25,8 +26,7 @@ import com.linecorp.armeria.common.annotation.UnstableApi;
 public final class DynamicEndpointGroupBuilder
         extends AbstractDynamicEndpointGroupBuilder {
 
-    @Nullable
-    private EndpointSelectionStrategy selectionStrategy;
+    private EndpointSelectionStrategy selectionStrategy = EndpointSelectionStrategy.weightedRoundRobin();
 
     DynamicEndpointGroupBuilder() {}
 
@@ -35,7 +35,7 @@ public final class DynamicEndpointGroupBuilder
      * If unspecified, {@link EndpointSelectionStrategy#weightedRoundRobin()} is used.
      */
     public DynamicEndpointGroupBuilder selectionStrategy(EndpointSelectionStrategy selectionStrategy) {
-        this.selectionStrategy = selectionStrategy;
+        this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
         return this;
     }
 
@@ -48,9 +48,6 @@ public final class DynamicEndpointGroupBuilder
      * Returns a newly created {@link DynamicEndpointGroup} with the properties configured so far.
      */
     public DynamicEndpointGroup build() {
-        if (selectionStrategy != null) {
-            return new DynamicEndpointGroup(selectionStrategy, shouldAllowEmptyEndpoints());
-        }
-        return new DynamicEndpointGroup(shouldAllowEmptyEndpoints());
+        return new DynamicEndpointGroup(selectionStrategy, shouldAllowEmptyEndpoints());
     }
 }

```

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:90]: Actually nevermind this comment. I guess we can leave since it is `protected`

**Comment by jrhee17** [comment]
I think I'm happy with these changes once my comments are addressed! Thanks in advance @doppany 🙇 🙏 🙇

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java]: Done

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java]: Done

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:90]: Yeah, I changed it to `protected` so that a subclass can still use it. :-)

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:183]: super nit; feel free to ignore this comment

I'm wondering why we're using guava's `Iterables` instead of getting size directly

**Review by jrhee17 — APPROVED** [review]
Thanks @doppany ! Congratulations on your first contribution 👍 🙇 👍

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:35]: 👍

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:25]: note: DNS cache PR introduced `AbstractDnsResolverBuilder` that conflicts with this builder.
`AbstractDynamicEndpointGroupBuilder` should extend `AbstractDnsResolverBuilder`.
https://github.com/line/armeria/pull/4133/files#diff-25f820493abf12e84fab771a6ef1ab4a6537fade6b9d31b1e95172d74be3f3e9R38

**Review by ikhoon — APPROVED** [review]
Thanks, @doppany and @trustin!

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:25]: I was confused.
`DnsEndpointGroupBuilder` needs to extend both `AbstractDynamicEndpointGroupBuilder` and `AbstractDnsResolverBuilder`.
Let me handle it in #4133 once this PR is merged.

**Comment by ikhoon** [comment]
Thanks for your first contribution, @doppany! ❤️ I hope to see your next PR. :-)

**Comment by trustin** [comment]
Added a follow-up PR which fixes the threading bug - https://github.com/line/armeria/commit/35c5c7633c112f08bd387f66b503822bdb96be3b

**Comment by doppany** [comment]
I'm just curious and asking. @trustin, is that commit https://github.com/line/armeria/commit/35c5c7633c112f08bd387f66b503822bdb96be3b related to this code  `endpointsLock.lock();`?

**Comment by trustin** [comment]
@doppany, there were two issues:

- `Iterables.size()` was redundant, as @jrhee17 said, because `endpoint` is a `List` and it already has `.size()` method.
- Any access to `this.endpoints` must be guarded by `endpointsLock` because `this.endpoints` itself isn't thread-safe and the update methods in `DynamicEndpointGroup` can be invoked from any threads.

**Comment by doppany** [comment]
@trustin Thank you for your kind explanation! :-) 🙏

**Pull request merged**
